### ISSUE TITLE
Event generator and mock server

### ISSUE THREAD

**Issue by AlexRuiz7** [wazuh/wazuh-kibana-app#3871]
**Description** 

We need to generate sample documents for elastic search using CI/CD compatible tools and compatible with our testing framework Cypress. We need to create a tool to generate events based on the templates built into the functional inventory.

- [x] Search for a mock server to use when developing the app. We need to mock the Wazuh API and ElasticSearch, though this should not be mandatory.

```mermaid
graph BT
    A(Inventory) --> B[Event Generator]
    B --> |Insert events using the API| C[Elastic / Splunk]
    C --> |Get indices using the API| D[UI]
    E[Wazuh API] --> |Mock Server| D
```

**Comment by AlexRuiz7** [comment]
Splunk provides an addon named Splunk Event Generator Utilty (EventGen):
> Provide sample data with your app or add-on using the Splunk Event Generator utility ([Eventgen](https://splunkbase.splunk.com/app/1924/)), along with an eventgen.conf file. Sample data helps demonstrate how your app or add-on functions during the code review. For more information, see the Eventgen app on Splunkbase.

**Comment by AlexRuiz7** [comment]
EventGen
---------

Some highlights.

> - Can be executed inside of Splunk (relying on a common event generation framework) as well as outside of Splunk
> - Event output can easily be directed to a Splunk input (modular inputs, HEC, etc.), a text file, or any REST endpoint in an extensible way
> - Easily configurable to make fake data look as real as possible, either by ordering events and token replacements by time of the day or by allowing generators to replay real data replacing current time by generating data exactly at the same time intervals as the original data


Mock Server
--------
[1]: https://www.mock-server.com/
[2]: https://learning.postman.com/docs/designing-and-developing-your-api/mocking-data/setting-up-mock/

| -   | [MockServer][1] | [Postman][2] | 
| --- | --------- | ------------ |
| OpenAPI support  | Yes | Yes   |

**Comment by AlexRuiz7** [comment]
## Task: Automatize the import of the Wazuh API Spec

1. Download the spec from Github using cURL.
2. Import the spec to the Mock Serve.
   1. Save the generated expectations to a file (`expectations.json`).
5. Make the expectations return the examples defined in the spec file.
  6. Read the `expectations.json` and sort the expectation by its operation ID (in memory).
  7. Read the `spec


* The `expectations.json` is a collection (array|list) of expectations objects, represented in a JSON format.

**Comment by AlexRuiz7** [comment]
## Task: Automatize the import of the Wazuh API Spec

1. Download the spec from Github using cURL.
2. Import the spec to the Mock Server.
   1. Save the generated expectations to a file (`expectations.json`).
5. Make the expectations return the examples defined in the spec file.
  6. Read the `expectations.json` and sort the expectation by its operation ID (in memory).
  7. Read the `spec.yaml` file


* The `expectations.json` is a collection (array|list) of expectations objects, represented in a JSON format.

**Comment by AlexRuiz7** [comment]
## Task: Automatize the import of the Wazuh API Spec

1. Download the spec from Github using cURL.
2. Import the spec to the Mock Server.
   1. Save the generated expectations to a file (`expectations.json`).
5. Make the expectations return the examples defined in the spec file.
   1. Read the `expectations.json` file and sort the expectation by its operation ID (in memory).
   7. Read the `spec.yaml` file, get the endpoints chunk and sort them by the operation ID (in memory)
   8. Match expectations and endpoints by key (operation ID) and for each, update the expectations response with the example in the spec.


* The `expectations.json` is a collection (array|list) of expectations objects, represented in a JSON format.

**Comment by AlexRuiz7** [comment]
## Status Update - 19/04/2022
Pulled the latest version of the MockServer Docker Image, which extends the verbosity for the cause of unmatched expectations.

The endpoints accepting arrays as query string parameters fail to match due to the following error:
  ```
  errors:

  1 error:
      - $: integer found, array expected
  ```

To determine if this an issue with our OpenAPI spec or a bug on the MockServer, I've tried using another OpenAPI spec that takes array on the query string parameters. The well known [Pet Store](https://petstore3.swagger.io/) spec has been used. This spec defines the following endpoint:
```yaml
  /pet/findByStatus:
  get:
    tags:
      - pet
    summary: Finds Pets by status
    description: Multiple status values can be provided with comma separated strings
    operationId: findPetsByStatus
    parameters:
      - name: status
        in: query
        description: Status values that need to be considered for filter
        required: false
        explode: true
        schema:
          type: string
          enum:
            - available
            - pending
            - sold
          default: available
    responses:
      '200':
        description: successful operation
        content:
          application/json:
            schema:
              type: array
              items:
                $ref: '#/components/schemas/Pet'
```

With this definition, the following request is accepted:

```
/pet/findByStatus?status=available&status=pending
```

but this is not:
```
/pet/findByStatus?status=available,pending
```

Following the documentation about the Query Parameters Serialization that was linked on the previous comment, we follow the **unexploded form** style. I modified the Pet Store spec to follow this style, and as a result, the request `/pet/findByStatus?status=available,pending` is now accepted.

This confirms that **the MockServer is able to work with arrays** as query string parameters, so we'll need to adjust our OpenAPI spec.

**Comment by AlexRuiz7** [comment]
Finally, we dropped MockServer due to bugs which won't allow us to import and use our OpenAPI spec correctly. Check the previous comments tagged as outdated for more information about this problem.

We ended up looking for a new mock server that would provide us the functionality we need: OpenAPI ready, highly configurable / scriptable, easy to install and use, and with good enough documentation.

After several tries, we have decided to use [Imposter](https://www.imposter.sh/), and we have been able to include in our new environments, which are hosted in this branch: [feature/3872-environments-docker](https://github.com/wazuh/wazuh-kibana-app/tree/feature/3872-environments-docker)

We are working on configuring and improving Imposter, the new Docker Environments and on an event generator that would allow us to populate Elastic / OpenSearch indices.
